File: wrapper/java/src/main/java/com/alpacacore/ModelDesc.java
// Copyright (c) Alpaca Core
// SPDX-License-Identifier: MIT
//
package com.alpacacore;

/// An descriptor for models.
/// @ingroup java-local
public class ModelDesc {
    /// Default constructor for an empty description.
    public ModelDesc() {}

    /// Constructor from components.
    public ModelDesc(String inferenceType, AssetInfo[] assets, String name) {
        this.inferenceType = inferenceType;
        this.assets = assets;
        this.name = name;
    }

    /// Inference type.
    /// Used as a router to the correct underlying inference library.
    public String inferenceType;

    /// Information for an individual asset.
    public static final class AssetInfo {
        public String path; ///< Path to the asset.
        public String tag; ///< Asset tag.
    }
    public AssetInfo[] assets; ///< Model assets.

    /// Name tag.
    /// This field is not used by the library in any way besides logs and may be helpful for debugging.
    public String name;
}

File: wrapper/swift/code/CxxAlpacaCore/ModelDesc.swift
/// Represents the information of an asset.
public class AssetInfo {
    /// The path to the asset.
    public var path: String = ""

    /// The tag of the asset. Some native model loaders may require it.
    public var tag: String = ""

    public init() {}

    public init(_ path:String, _ tag:String) {
        self.path = path
        self.tag = tag
    }
}

/// The description of the model used by the native `ModelLoader`.
public class ModelDesc {
    /// The type of the inference used to select the appropriate model loader.
    public var inferenceType: String = ""

    /// Name tag. Not used by the library in any way besides logs and may be helpful for debugging
    public var name: String = ""

    /// The list of assets required by the model.
    public var assets: [AssetInfo]

    public init() {
        self.assets = []
    }

    public init(_ inferenceType:String = "", _ name:String = "", _ assets:[AssetInfo] = []) {
        self.inferenceType = inferenceType
        self.name = name
        self.assets = assets
    }
}
